x/trigger/keeper/event_detector.go

```diff
@@ -0,0 +1,72 @@
+package keeper
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/provenance-io/provenance/x/trigger/types"
+)
+
+// DetectBlockEvents Detects triggers that have been activated by their corresponding events.
+func (k Keeper) DetectBlockEvents(ctx sdk.Context) {
+	triggers := k.detectTransactionEvents(ctx)
+	triggers = append(triggers, k.detectBlockHeightEvents(ctx)...)
+	triggers = append(triggers, k.detectTimeEvents(ctx)...)
+
+	for _, trigger := range triggers {
+		k.UnregisterTrigger(ctx, trigger)
+		k.QueueTrigger(ctx, trigger)
+	}
+}
+
+// DetectTransactionEvents Detects triggers that have been activated by transaction events.
+func (k Keeper) detectTransactionEvents(ctx sdk.Context) (triggers []types.Trigger) {
+	detectedTriggers := map[uint64]bool{}
+	for _, event := range ctx.EventManager().GetABCIEventHistory() {
+		matched := k.getMatchingTriggers(ctx, event.GetType(), func(trigger types.Trigger, triggerEvent types.TriggerEventI) bool {
+			if _, isDetected := detectedTriggers[trigger.Id]; isDetected {
+				return false
+			}
+			txEvent := triggerEvent.(*types.TransactionEvent)
+			detected := txEvent.Equals(event)
+			detectedTriggers[trigger.Id] = detected
+			return detected
+		})
+		triggers = append(triggers, matched...)
+	}
+	return
+}
+
+// DetectBlockHeightEvents Detects triggers that have been activated by block height events.
+func (k Keeper) detectBlockHeightEvents(ctx sdk.Context) (triggers []types.Trigger) {
+	triggers = k.getMatchingTriggers(ctx, types.BlockHeightPrefix, func(_ types.Trigger, triggerEvent types.TriggerEventI) bool {
+		blockHeightEvent := triggerEvent.(*types.BlockHeightEvent)
+		return ctx.BlockHeight() >= int64(blockHeightEvent.GetBlockHeight())
+	})
```

Comment:
Here too, I'm not sure how sustainable this is.

It'd be better to put the height in the key, e.g. `<prefix><height><trigger id>`. Then, the iterator would start at `<prefix>` and end at `<prefix><current height><max uint64>`. No need to process any of the others since they happen at later blocks.